### DISCUSSION TITLE
Dynamically setting `publishDir` `mode`

### DISCUSSION THREAD

**Discussion by jrm5100** [nextflow-io/nextflow#2403] [Q&A] [answered]
Do closures only work if the entire directive is specified as a string?

Based on the documentation [for dynamic directives](https://www.nextflow.io/docs/latest/process.html#dynamic-directives) I was hoping a closure would work for `mode` under the `publishDir` directive, but I get:

```
Invalid method invocation `setMode` with arguments:  (Script_a917c133$_runScript_closure1$_closure2) on PublishDir type
```

Here is my attempt:

```
publishDir params.output, mode: {move ? "move" : "symlink"}
```

with a `val move` that gets passed `true` or `false` under `input`.

I also tried putting the entire directive in a closure, but the module (as I'd expect) fails to compile.

**Comment by manuelesimi** [comment, marked as answer]
I'm not sure you can use a closure there. However, you can go with plain Groovy and use this:

`publishDir params.output, mode: "${move ? 'move' : 'symlink'}"`

**Reply by jrm5100** [reply]
Thanks!

**Comment by mjhipp** [comment]
This suggested change should also allow you to use actual closures there: https://github.com/nextflow-io/nextflow/pull/2432
